lib/dotcom_web/components/mode_icons.ex

```diff
@@ -0,0 +1,154 @@
+  use Phoenix.Component
+
+  import DotcomWeb.Components.{Badges, Svg}
+  import MbtaMetro.Components.{Badge, Icon}
+
+  attr :route_name, :string, required: true
+
+  attr :type, :atom,
+    values: [
+      :subway,
+      :commuter_rail,
+      :bus,
+      :ferry,
+      :logan_express,
+      :massport_shuttle,
+      :orange_line,
+      :red_line,
+      :green_line,
+      :green_line_b,
+      :green_line_c,
+      :green_line_d,
+      :green_line_e,
+      :blue_line,
+      :mattapan_line
+    ],
+    required: true
+
+  attr :class, :string, default: ""
+
+  def mode_icon(%{type: :bus} = assigns) do
+    ~H"""
+    <.rect_icon class={"bg-yellow-500 #{@class}"}><%= @route_name %></.rect_icon>
+    """
+  end
+
+  def mode_icon(%{type: :silver_line} = assigns) do
+    ~H"""
+    <.rect_icon class={"bg-silver-500 text-white #{@class}"}><%= @route_name %></.rect_icon>
+    """
+  end
+
+  def mode_icon(%{type: :massport_shuttle} = assigns) do
+    ~H"""
+    <.circle_icon class={"bg-navy-500 text-md #{@class}"}><%= @route_name %></.circle_icon>
+    """
+  end
+
+  def mode_icon(%{type: :blue_line} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-blue-500 #{@class}"}>BL</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :green_line} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-green-500 #{@class}"}>GL</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :green_line_b} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-green-500 #{@class}"}>B</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :green_line_c} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-green-500 #{@class}"}>C</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :green_line_d} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-green-500 #{@class}"}>D</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :green_line_e} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-green-500 #{@class}"}>E</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :orange_line} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-orange-500 #{@class}"}>OL</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :red_line} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-red-500 #{@class}"}>RL</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{type: :mattapan_line} = assigns) do
+    ~H"""
+    <.subway_icon class={"bg-red-500 #{@class}"}>M</.subway_icon>
+    """
+  end
+
+  def mode_icon(%{route_name: route_name, type: :logan_express} = assigns) do
+    fill_color_class =
+      case route_name do
+        "BB" -> "fill-logan-express-bb"
+        "BT" -> "fill-logan-express-bt"
+        "DV" -> "fill-logan-express-dv"
+        "WO" -> "fill-logan-express-wo"
+        _ -> ""
+      end
+
+    assigns = assign(assigns, :fill_color_class, fill_color_class)
+
+    ~H"""
+    <.icon class={"w-6 #{@fill_color_class} #{@class}"} type="icon-svg" name="icon-logan-express" />
+    """
+  end
```

Comment:
```
// This could and probably should live where routes are defined
@logan_express_routes ~w[BB BT DV WO]

def logan_express_routes, do: @logan_express_routes

// Then in this module

@logan_express_routes Routes.logan_express_routes()

def mode_icon(%{route_name: route_name, type: :logan_express} = assigns) when route_name in @logan_express_routes do
  ~H"""
  <.icon class={"w-6 fill-logan-express-#{String.downcase(@route_name}"} ... />  
  """
end
```

Makes it much easier to test because you can get a list of routes.

![Screenshot Capture - 2024-11-06 - 16-53-29](https://github.com/user-attachments/assets/1cdc0a19-96e0-4ef6-9d48-ca91b6880ac4)

Reply:
I like that, especially because then there's one (maybe two) places that need to be updated if we ever get another Logan Express route.

The one thing that's throwing me off is that tailwind seems to only reliably recognize `fill-logan-express-#{String.downcase(@route_name}` as a valid utility class when `fill-logan-express-bb` appears explicitly in the code - the string interpolation thing sometimes worked for me and sometimes didn't (I ran into this problem when I tried string-interpolating the subway line colors into `bg-#{color}-500` as well). Not knowing 100% about how tailwind works under the hood, I guessed that it was scanning the code for usages of the various utility classes, and only generating the ones that it thought were needed, which left out ones that were programmatically generated, but maybe that's not actually how it works?

Do you know how tailwind does this? And/or how to convince tailwind to just generate the utility classes we need?

Reply:
https://github.com/mbta/mbta_metro/blob/main/assets/js/styles/safelist.js